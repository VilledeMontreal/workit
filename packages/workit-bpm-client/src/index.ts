/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

// IOC
export * from './config/constants/identifiers';
import './config/container';

export * from './camundaBpmClient';
export * from './bpmClient';
export * from './bpmLogger';
export * from './camundaBpmClient';
export * from './camundaMapperProperties';
export * from './logger';
export * from './camundaMessage';

export * from './utils/utils';
export * from './variables';
