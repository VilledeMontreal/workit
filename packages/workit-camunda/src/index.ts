// Copyright (c) Ville de Montreal. All rights reserved.
// Licensed under the MIT license.
// See LICENSE file in the project root for full license information.

import './config/ioc';

export * from './config/constants';

// IOC
export * from './config/constants/identifiers';
export * from './config/constants/tag';
export * from './config/ioc';
export * from './models/IoC';

// Camunda specific
// // specs
export * from './models/camunda/specs/camundaClient';
export * from './models/camunda/specs/camundaConfig';
export * from './models/camunda/specs/subscriptionOptions';
export * from './models/camunda/specs/payload';
// // General
export * from './models/camunda/utils';
export * from './models/camunda/variables';
export * from './models/camunda/camundaBpmClient';
export * from './models/camunda/camundaMapperProperties';
export * from './models/camunda/camundaMessage';
export * from './models/camunda/logger';

// Zeebe specific
// // specs
export * from './models/zeebe/specs/payload';
// export * from './models/zeebe/specs/workflowDeployResponse';
export * from './models/zeebe/specs/zeebeClientOptions';
export * from './models/zeebe/specs/zeebeOptions';
export * from './models/zeebe/specs/zeebeWorkerOptions';
export * from './models/zeebe/specs/logLevel';
// // General
export * from './models/zeebe/zeebeClient';
export * from './models/zeebe/zeebeMapperProperties';
export * from './models/zeebe/zeebeMessage';

// Client lib
export * from './models/camunda-n-mq/specs/incidentException';
export * from './models/camunda-n-mq/specs/failureException';
export * from './models/camunda-n-mq/specs/camundaService';
export * from './models/camunda-n-mq/specs/client';
export * from './models/camunda-n-mq/specs/workflowClient';
export * from './models/camunda-n-mq/specs/message';
export * from './models/camunda-n-mq/specs/properties';
export * from './models/camunda-n-mq/specs/propertiesBase';
export * from './models/camunda-n-mq/specs/customHeaders';
export * from './models/camunda-n-mq/specs/createWorkflowInstance';
export * from './models/camunda-n-mq/specs/createWorkflowInstanceResponse';
export * from './models/camunda-n-mq/specs/deployWorkflowResponse';
export * from './models/camunda-n-mq/specs/publishMessage';
export * from './models/camunda-n-mq/specs/updateWorkflowVariables';
export * from './models/camunda-n-mq/specs/workflow';
export * from './models/camunda-n-mq/specs/workflowResponse';
export * from './models/camunda-n-mq/client';
export * from './models/camunda-n-mq/clientManager';
export * from './utils/utils';

// Core
export * from './models/core/processHandler/specs/processHandlerConfig';
export * from './models/core/specs/process';
export * from './models/core/specs/task';
export * from './models/core/specs/failureStrategy';
export * from './models/core/specs/successStrategy';
export * from './models/core/worker';
export * from './models/core/specs/taskBase';
export * from './models/core/specs/interceptor';

// Core instrumentation
export * from './models/core/instrumentations/specs/camundaClientInstrumentation';
export * from './models/core/instrumentations/specs/camundaClientTracer';
export * from './models/core/instrumentations/specs/instrumentation';
export * from './models/core/instrumentations/camundaClientTracer';
export * from './models/core/instrumentations/enums/apm';
