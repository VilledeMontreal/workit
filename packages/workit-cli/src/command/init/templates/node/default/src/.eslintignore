node_modules
dist
.vscode
coverage
*lock*
*.md
tests