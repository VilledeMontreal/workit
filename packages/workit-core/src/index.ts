/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

export * from './config/container';
export * from './config/constants/identifiers';
export * from './config/constants';

export * from './processHandler/simpleCamundaProcessHandler';
export * from './interceptors';

export * from './specs/taskBase';

export * from './strategies/FailureStrategySimple';
export * from './strategies/SuccessStrategySimple';

export * from './IoC';

// Proxy
export * from './proxyFactory';
export * from './proxyObserver';

export * from './worker';

export * from './utils/utils';
