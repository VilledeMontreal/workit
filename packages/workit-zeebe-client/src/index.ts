/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

// IOC
export * from './config/constants/identifiers';
import './config/container';

export * from './zeebeClient';
export * from './zeebeMapperProperties';
export * from './zeebeMessage';
