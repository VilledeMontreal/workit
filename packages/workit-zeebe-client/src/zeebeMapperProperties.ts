/*
 * Copyright (c) 2020 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */
import { IEmptyPayload, IPayload, IWorkflowProps } from 'workit-types';

export class ZeebeMapperProperties {
  public static map<TVariables = unknown, TProps = unknown>(obj: IPayload<TVariables, TProps>): IWorkflowProps<TProps> {
    let businessKey;
    if (obj.variables || (obj.variables as any).businessKey) {
      businessKey = (obj.variables as any).businessKey;
    }

    return {
      businessKey, // retro compatibility with bpmn workflow engine
      activityId: obj.elementId,
      processInstanceId: obj.elementInstanceKey,
      workflowDefinitionVersion: obj.workflowDefinitionVersion,
      workflowInstanceKey: obj.workflowInstanceKey,
      workflowKey: obj.workflowKey,
      bpmnProcessId: obj.bpmnProcessId,
      customHeaders: { ...obj.customHeaders },
      jobKey: obj.key,
      retries: obj.retries,
      topicName: obj.type,
      workerId: obj.worker,
      lockExpirationTime: new Date(Number(obj.deadline))
    };
  }

  public static unmap<TProps = unknown>(props: IWorkflowProps<TProps>): IEmptyPayload<TProps> {
    return {
      elementId: props.activityId,
      elementInstanceKey: props.processInstanceId,
      workflowKey: props.workflowKey,
      workflowInstanceKey: props.workflowInstanceKey,
      workflowDefinitionVersion: props.workflowDefinitionVersion,
      bpmnProcessId: props.bpmnProcessId,
      retries: props.retries,
      worker: props.workerId,
      deadline: props.lockExpirationTime.getTime().toString(),
      customHeaders: { ...props.customHeaders },
      key: props.jobKey,
      type: props.topicName
    };
  }
}
