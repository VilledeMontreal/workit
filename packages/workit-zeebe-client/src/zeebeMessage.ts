/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { getVariablesWhenChanged } from 'workit-core';
import { FailureException, ICamundaService, IMessage, IPayload, IWorkflowProps } from 'workit-types';
// FIXME:dist folder....
import { CompleteFn } from 'zeebe-node/dist/lib/interfaces';
import { ZeebeMapperProperties } from './zeebeMapperProperties';

export class ZeebeMessage {
  public static wrap<TVariables, TProps>(
    payload: IPayload<TVariables, TProps>,
    complete: CompleteFn<TVariables>
  ): [IMessage<TVariables, IWorkflowProps<TProps>>, ICamundaService] {
    const properties = ZeebeMapperProperties.map(payload);
    return [
      {
        body: payload.variables,
        properties
      },
      {
        hasBeenThreated: false,
        async ack(message: IMessage<Partial<TVariables> | undefined, IWorkflowProps<unknown>>) {
          if (this.hasBeenThreated) {
            return Promise.resolve();
          }

          // TODO: change any to real type body
          const vars = getVariablesWhenChanged<any>(message, msg => ZeebeMessage.unwrap(msg));

          this.hasBeenThreated = await complete.success(vars.variables);
        },
        nack(error: FailureException) {
          if (this.hasBeenThreated) {
            return Promise.resolve();
          }
          const retries = error.retries;
          // TODO: check if zeebe-node made the type correction
          this.hasBeenThreated = (complete.failure(error.message, retries) as unknown) as boolean;
          return Promise.resolve();
        }
      }
    ];
  }

  /**
   * Shallow copy
   */
  public static unwrap<TVariables, TProps>(
    message: IMessage<TVariables, IWorkflowProps<TProps>>
  ): IPayload<TVariables, TProps> {
    const emptyPayload = ZeebeMapperProperties.unmap<TProps>(message.properties);
    (emptyPayload as IPayload<TVariables, TProps>).variables = message.body;
    return emptyPayload as IPayload<TVariables, TProps>;
  }
}
