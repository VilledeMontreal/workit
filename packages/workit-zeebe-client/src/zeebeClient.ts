/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */
import { optional } from 'inversify';
import {
  ICamundaService,
  IClient,
  ICreateWorkflowInstance,
  ICreateWorkflowInstanceResponse,
  IDeployWorkflowResponse,
  IMessage,
  IPagination,
  IPaginationOptions,
  IPayload,
  IPublishMessage,
  IUpdateWorkflowRetry,
  IUpdateWorkflowVariables,
  IWorkflow,
  IWorkflowClient,
  IWorkflowDefinition,
  IWorkflowDefinitionKey,
  IWorkflowDefinitionRequest,
  IWorkflowOptions,
  IWorkflowProcessIdDefinition,
  IWorkflowProps,
  IZeebeOptions
} from 'workit-types';
import { Configs, IAPIConfig as IElasticExporterConfig, ZBElasticClient } from 'zeebe-elasticsearch-client';
import { ZBClient, ZBWorker } from 'zeebe-node';
// FIXME: dist folder
import { CompleteFn } from 'zeebe-node/dist/lib/interfaces';
import { PaginationUtils } from './utils/paginationUtils';
import { ZeebeMessage } from './zeebeMessage';

export class ZeebeClient<TVariables = unknown, TProps = unknown, RVariables = TVariables>
  implements IClient<ICamundaService>, IWorkflowClient {
  private readonly _client: ZBClient;
  private readonly _exporterClient: ZBElasticClient;
  private _worker: ZBWorker<TVariables, TProps, RVariables> | undefined;
  private readonly _config: IZeebeOptions;
  private readonly _exporterConfig: Partial<IElasticExporterConfig> | undefined;
  constructor(config: IZeebeOptions, client?: ZBClient, exporterConfig?: Partial<IElasticExporterConfig>);
  constructor(config: IZeebeOptions, exporterConfig?: Partial<IElasticExporterConfig>);
  constructor(
    config: IZeebeOptions,
    @optional() client?: ZBClient | Partial<IElasticExporterConfig>,
    @optional() exporterConfig?: Partial<IElasticExporterConfig>
  ) {
    this._config = config;

    if (client && (client as Partial<IElasticExporterConfig>).url) {
      this._exporterConfig = client as Partial<IElasticExporterConfig>;
      this._client = new ZBClient(config.baseUrl, config);
    } else {
      this._client = (client || new ZBClient(config.baseUrl, config)) as ZBClient;
      this._exporterConfig = exporterConfig;
    }

    if (!exporterConfig) {
      // tslint:disable-next-line: no-console
      console.log(
        "warning: no exporterConfig has been provided to Zeebe. getWorkflow and getWorkflows methods won't work. "
      );
    }
    this._exporterClient = new ZBElasticClient(new Configs(exporterConfig));
  }
  public subscribe(
    onMessageReceived: (
      message: IMessage<TVariables, IWorkflowProps<TProps>>,
      service: ICamundaService
    ) => Promise<void>
  ): Promise<void> {
    this._worker = this._client.createWorker(
      this._config.workerId || 'some-random-id',
      this._config.topicName,
      async (payload: IPayload<TVariables, TProps>, complete: CompleteFn<TVariables>) => {
        const [message, service] = ZeebeMessage.wrap<TVariables, TProps>(payload, complete);
        await onMessageReceived(message, service);
      },
      this._config
    );
    return Promise.resolve();
  }
  public async deployWorkflow(bpmnPath: string): Promise<IDeployWorkflowResponse> {
    const result = await this._client.deployWorkflow(bpmnPath);
    return {
      workflows: result.workflows,
      key: result.key.toString() // TODO: interface say number but it return string, need to PR to zeebe-node
    };
  }
  public async getWorkflows(options?: Partial<IWorkflowOptions & IPaginationOptions>): Promise<IPagination<IWorkflow>> {
    this._validateExporterConfig();
    const params = { _source_excludes: 'bpmnXml' };
    const workflowOptions = { params };
    workflowOptions.params = PaginationUtils.setElasticPaginationParams(params, options);
    const criteria = this._setWorkflowCriteria(options);
    const result = await this._exporterClient.getWorkflows(criteria, workflowOptions);
    const elasticResult = result.data.hits;
    const data = elasticResult.hits.map(doc => {
      const workflow = doc._source;
      return {
        bpmnProcessId: workflow.bpmnProcessId,
        version: workflow.version,
        workflowKey: workflow.key.toString(),
        resourceName: workflow.resourceName
      };
    });

    return {
      paging: PaginationUtils.getPagingFromOptions(elasticResult.total, options),
      items: data
    };
  }
  public async getWorkflow(payload: IWorkflowDefinitionRequest): Promise<IWorkflowDefinition> {
    this._validateExporterConfig();
    this._validateObject();
    const key = Number((payload as IWorkflowDefinitionKey).workflowKey);
    const response = await this._exporterClient.getWorkflows({
      key: !isNaN(key) ? key : undefined,
      bpmnProcessId: (payload as IWorkflowProcessIdDefinition).bpmnProcessId,
      version: (payload as IWorkflowProcessIdDefinition).version,
      latestVersion: typeof (payload as IWorkflowProcessIdDefinition).version !== 'number'
    });

    const data = response.data.hits.hits[0];

    if (!data) {
      throw new Error('Not found');
    }

    const doc = data._source;

    return {
      version: doc.version,
      resourceName: doc.resourceName,
      bpmnXml: doc.bpmnXml,
      workflowKey: doc.key.toString(),
      bpmnProcessId: doc.bpmnProcessId
    };
  }
  public updateVariables<T = unknown>(model: IUpdateWorkflowVariables<Partial<T>>): Promise<void> {
    return this._client.setVariables<T>({
      elementInstanceKey: model.processInstanceId,
      variables: model.variables,
      local: !!model.local
    });
  }
  public updateJobRetries(payload: IUpdateWorkflowRetry): Promise<void> {
    return this._client.updateJobRetries(payload);
  }
  /**
   * Publish a message event
   * If you don't specify correlationKey, message will be treated as message start event
   */
  public publishMessage<T = unknown>(payload: IPublishMessage<T, string>): Promise<void> {
    return this._client.publishMessage({
      correlationKey: payload.correlation || '__MESSAGE_START_EVENT__',
      variables: payload.variables,
      messageId: payload.messageId,
      timeToLive: payload.timeToLive,
      name: payload.name
    });
  }
  public createWorkflowInstance<T = unknown>(
    model: ICreateWorkflowInstance<T>
  ): Promise<ICreateWorkflowInstanceResponse> {
    if (!model.version) {
      return this._client.createWorkflowInstance<T>(model.bpmnProcessId, model.variables);
    }
    return this._client.createWorkflowInstance<T>(model as Required<ICreateWorkflowInstance<T>>);
  }
  public cancelWorkflowInstance(instance: string): Promise<void> {
    this._validateNumber(instance);
    return this._client.cancelWorkflowInstance(instance as any); // TODO: will be fixed https://github.com/zeebe-io/zeebe/issues/2680
  }
  public resolveIncident(incidentKey: string): Promise<void> {
    return this._client.resolveIncident(incidentKey);
  }
  public unsubscribe(): Promise<void> {
    if (this._worker) {
      return this._worker.close();
    }
    return Promise.resolve();
  }

  private _validateExporterConfig() {
    if (!this._exporterConfig) {
      throw new Error(`
      Please, refer to the warning when you instiate this class. You must pass exporterConfig to the Ctor in order to use this method.
      For now, we are only compatible with elastic indexes that operate produce. If you use different indexes or different exporter, please raise an issue.
      `);
    }
  }
  private _validateObject() {
    if (!this._exporterConfig) {
      throw new Error(`
        Object passed to the method can't be undefined
      `);
    }
  }
  private _validateNumber(variable: string) {
    const value = Number(variable);
    if (!Number.isInteger(value)) {
      throw new Error(`
      workflowInstanceKey value is malformed
      `);
    }
  }

  private _setWorkflowCriteria(options?: Partial<IWorkflowOptions & IPaginationOptions>) {
    if (!options) {
      return {};
    }
    return { bpmnProcessId: options.bpmnProcessId };
  }
}
