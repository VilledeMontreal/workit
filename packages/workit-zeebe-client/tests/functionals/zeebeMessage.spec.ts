/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { ZeebeMessage } from '../../src/zeebeMessage';

describe('zeebeMessage', () => {
  it('unmap', () => {
    const message = {
      body: { hello: 'world' },
      properties: {
        workflowInstanceKey: '38963',
        bpmnProcessId: 'test-process',
        workflowDefinitionVersion: 4,
        workflowKey: '8806',
        activityId: 'ServiceTask_0xdwuw7',
        elementInstanceKey: '38967',
        customHeaders: { jwt: 'jwt fake' },
        jobKey: '38968',
        processInstanceId: '38963',
        retries: 1,
        lockExpirationTime: new Date(1562269537659),
        topicName: 'topic_demo',
        workerId: 'demo',
      },
    };
    const zeebeObject = ZeebeMessage.unwrap(message);
    expect(zeebeObject).toMatchSnapshot();
  });
});
