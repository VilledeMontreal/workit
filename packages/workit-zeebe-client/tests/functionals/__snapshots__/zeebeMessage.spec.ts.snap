// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zeebeMessage unmap 1`] = `
Object {
  "bpmnProcessId": "test-process",
  "customHeaders": Object {
    "jwt": "jwt fake",
  },
  "deadline": "1562269537659",
  "elementId": "ServiceTask_0xdwuw7",
  "elementInstanceKey": "38963",
  "key": "38968",
  "retries": 1,
  "type": "topic_demo",
  "variables": Object {
    "hello": "world",
  },
  "worker": "demo",
  "workflowDefinitionVersion": 4,
  "workflowInstanceKey": "38963",
  "workflowKey": "8806",
}
`;
